runtime/src/tokio/telemetry.rs

```diff
@@ -26,7 +32,35 @@ pub fn init(context: Context, level: Level, metrics: Option<SocketAddr>, traces:
     if let Some(cfg) = metrics {
         context
             .with_label("metrics")
-            .spawn(move |context| async move { metrics::server::serve(context, cfg).await });
+            .spawn(move |context| async move {
+                // Create a listener for the metrics server
+                let listener = context
+                    .bind(cfg)
+                    .await
+                    .expect("Could not bind to metrics address");
+
+                // Create a router for the metrics server
+                let app = Router::new()
+                    .route(
+                        "/metrics",
+                        get(|Extension(ctx): Extension<Context>| async move {
+                            Response::builder()
+                                .status(StatusCode::OK)
+                                .header(header::CONTENT_TYPE, "text/plain; version=0.0.4")
+                                .body(Body::from(ctx.encode()))
+                                .expect("Failed to create response")
+                        }),
+                    )
+                    .layer(Extension(context));
+
+                // Serve the metrics over HTTP
+                //
+                // `serve` will spawn its own tasks using `tokio`. These will not be tracked
+                // like metrics spawned by `context`.
+                serve(listener, app.into_make_service())
```

Comment:
TODO: We should use an external listener tuned for metrics (rather than abusing the same network implementation we use for p2p).